### PR TITLE
[BugFix] cherry-pick RleEncoderV2 related code from ORC's github

### PR DESCRIPTION
**Pull request by cbcbq**
## Why I'm doing:
image

企业微信截图_0c74d37d-25a4-4556-8120-9783a7cd19da

There's a bug when writting ORC file using patched-base to encode integers under following conditions:

1. option.min (the base in patched-base encoding) is a negative number
2. the bit width of abs(option.min) is exactly a multiple of 8

in that situation, the decode code in Figure 2 will mistakenly interprets the MSB as the sign bit, leading to the use of an incorrect base and resulting in incorrect outcomes.

be-ut before this pr:
企业微信截图_4630d9a5-1e95-4302-92bd-3dd9fd01d8cd

## What I'm doing:

I found that the [ORC codes](https://github.com/apache/orc) doesn't have this bug.
So I just merged the RleEncoderV2.h/cpp and little related chages into SR entirely.

Fixes #issue

## What type of PR is this:

- [x] BugFix
- [ ] Feature
- [ ] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

Does this PR entail a change in behavior?

- [ ] Yes, this PR will result in a change in behavior.
- [x] No, this PR will not result in a change in behavior.

If yes, please specify the type of change:

- [ ] Interface/UI changes: syntax, type conversion, expression evaluation, display information
- [ ] Parameter changes: default values, similar parameters but with different default values
- [ ] Policy changes: use new policy to replace old one, functionality automatically enabled
- [ ] Feature removed
- [ ] Miscellaneous: upgrade & downgrade compatibility, etc.

## Checklist:

- [ ] I have added test cases for my bug fix or my new feature
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function
- [ ] This is a backport pr

## Bugfix cherry-pick branch check:
- [x] I have checked the version labels which the pr will be auto-backported to the target branch
  - [x] 3.3
  - [x] 3.2
  - [x] 3.1
  - [x] 3.0
  - [ ] 2.5

### CODE REVIEW

**Comment by cbcbq** [comment]
run starrocks_be_unittest

**Comment by cbcbq** [comment]
@Mergifyio refresh

**Comment by cbcbq** [comment]
@imay could you please assign someone to review this?

**Comment by Smith-Cruise** [comment]
can you paste orc's bugfix url from apache orc

**Review by dirtysalt — DISMISSED** [review]

**Review by zombee0 — COMMENTED** [review]
could you add some ut?

**Comment by cbcbq** [comment]
> can you paste orc's bugfix url in apache orc?

in fact it's just not an ORC's bug, the issue codes can't be found in ORC's github repo.
it might originate from very early versions of SR or Doris I think

**Comment by cbcbq** [comment]
> could you add some ut?

Ok done.
before pr:
image

after:
image

**Review by Smith-Cruise — APPROVED** [review]

File: be/src/formats/orc/apache-orc/c++/src/RleEncoderV2.cc [hunk 1]
```diff
@@ -583,9 +556,7 @@ void RleEncoderV2::writePatchedBasedValues(EncodingOption& option) {
     // find the number of bytes required for base and shift it by 5 bits
     // to accommodate patch width. The additional bit is used to store the sign
     // of the base value.
-    uint32_t baseBitsNum = findClosestNumBits(option.min);
```
Comment (Smith-Cruise) [inline be/src/formats/orc/apache-orc/c++/src/RleEncoderV2.cc:586]: wired, sr is copying orc's code directly, but this line code looks like never existed

**Review by Smith-Cruise — APPROVED** [review]

**Review by dirtysalt — APPROVED** [review]

**Review by Smith-Cruise — APPROVED** [review]

**Review by zombee0 — APPROVED** [review]

**Review by Youngwb — APPROVED** [review]

**Comment by Smith-Cruise** [comment]
@Mergify backport branch-3.3 branch-3.2 branch-3.1 branch-3.0

**Pull request merged**